facility/create-facility-with-file-upload.js

```diff
@@ -0,0 +1,141 @@
+/*
+    This example demonstrates how to create facility from uploaded file.
+    
+    It uses 2-legged authentication.
+
+    NOTE - the example uses API which is NOT SUPPORTED at the moment:
+        POST https://developer.api.autodesk.com/tandem/v1/groups/:groupId/twins
+        POST https://developer.api.autodesk.com/tandem/v1/twins/:facilityId/template
+        POST https://developer.api.autodesk.com/tandem/v1/twins/:facilityId/model
+        POST https://developer.api.autodesk.com/tandem/v1/twins/:facilityId/import
+        GET https://developer.api.autodesk.com/tandem/v1/models/:modelId/props
+*/
+import path from 'path';
+
+import { createToken } from '../common/auth.js';
+import { TandemClient } from '../common/tandemClient.js';
+import { ModelState, QC, readBinary, readJSON } from '../common/utils.js';
+
+// update values below according to your environment
+const APS_CLIENT_ID = 'YOUR_CLIENT_ID';
+const APS_CLIENT_SECRET = 'YOUR_CLIENT_SECRET';
+const GROUP_URN = 'YOUR_GROUP_URN'; // account where the facility will be created
+const INPUT_MODEL_NAME = 'C:/Tem/racbasicsampleproject.rvt'; // local file to create facility from
+
+async function main() {
+    // STEP 1 - obtain token. The sample uses 2-legged token but it would also work with 3-legged token
+    // assuming that user has Manage permission which allows facility creation
+    // note that it requires data:write scope
+    const token = await createToken(APS_CLIENT_ID,
+        APS_CLIENT_SECRET, 'data:read data:write');
+    const client = new TandemClient(() => {
+        return token;
+    });
+    
+    // STEP 2 - create facility
+    const facilityInputs = {
+        settings: {
+            props: {
+                'Identity Data': {
+                    'Building Name': 'Basic Sample',
+                    'Address': '1 Market St. 4th Floor, San Francisco, CA, 94105',
+                    'Owner': 'Jan Liska',
+                    'Project Name': 'Basic Sample'
+                }
+            }
+        }
+    };
+    const facilityResult = await client.createFacility(GROUP_URN, facilityInputs);
+    const facilityId = facilityResult[QC.Key];
+
+    console.log(`new facility: ${facilityId}`);
+    // STEP 3 - read & apply facility template - we use data downloaded from Tandem app server
+    const template = await readJSON('./data/facilityTemplate.json');
+
+    await client.applyFacilityTemplate(facilityId, template);
+    // STEP 4 - create link to upload file to S3
+    console.log(`uploading file: ${INPUT_MODEL_NAME}`);
+    const uploadLink = await client.createUploadLink(facilityId, path.basename(INPUT_MODEL_NAME));
+    // STEP 5 - upload file
+    const fileContent = await readBinary(INPUT_MODEL_NAME);
```

Comment:
This might be problematic for larger files, where you would rather stream them directly, but that's probably ok for a sample code.

Reply:
Added comment